baybe/utils/dataframe.py

```diff
@@ -361,6 +361,50 @@ def df_uncorrelated_features(
     return data
 
 
+def df_add_noise_to_degenerate_rows(
+    df: pd.DataFrame, noise_level: float = 0.001
+) -> pd.DataFrame:
+    """Add noise to degenerate rows to being able to numerically distinguish them.
+
+    Note that the dataframe is changed in-place and also returned.
+
+    Args:
+        df: The dataframe to be modified.
+        noise_level: The magnitude of noise relative to the min-max range of
+            values for each column.
+
+    Returns:
+        The modified dataframe.
+
+    Raises:
+        TypeError: If the provided dataframe has non-numerical content.
+    """
+    # Assert that the input is purely numerical
+    if any(df[col].dtype.kind not in "iufb" for col in df.columns):
+        raise TypeError(
+            f"'{df_add_noise_to_degenerate_rows.__name__}' only supports purely"
+            f"numerical dataframes."
+        )
+
+    # Find the min-max range for each column. Constant columns will be assigned a range
+    # of 1 as fallback as otherwise they would be left untouched
+    column_ranges = df.max() - df.min()
+    column_ranges = column_ranges.replace(0, 1)
+
+    # Find degenerate rows and generate noise
+    degen_rows = df.duplicated(keep=False)
+    noise = np.random.normal(
+        -noise_level, noise_level, size=(degen_rows.sum(), df.shape[1])
+    )
+    noise_df = pd.DataFrame(noise, columns=df.columns, index=df.index[degen_rows])
+
+    # Scale noise by column ranges and add it to the original dataframe
+    noise_df = noise_df * column_ranges
+    df.loc[degen_rows] += noise_df
+
+    return df
```

Comment:
I am a bit surprised about the details of the implementation. Here the things I noticed, perhaps you can comment:
* Probably you wanted using `np.random.uniform` in some way to properly reflect the random percentage of noise, instead of setting loc and scale of a Gaussian?
* You do a weird mix (at least in my opinion) of dataframe/array mix in your computation. Seems to work, but I'm always a bit skeptical with that. In particular
  *  Setting both the index of `noise_df` and then addressing `df.loc[degen_rows]` is somehow redundant
  * Matrix product implemented via product of dataframe and pandas series ... yikes 😄 is this intentional? Perhaps this is a common thing but for a numpy-fan like me it seems very strange

Reply:
the uniform just directly implements the idea of using a noise thats proportional to the minmax-range. With normal noise you can scale it too according to that range, but its not quite as clear. What should sigma be if I want noise on the order of 0.01 of the minmax scale?? Just plusminus 0.01*minmax range seems more comprehendable to me. Overall it doesnt matter but since you raised the point do you see any arguments for/against or do we not need to discuss this any further? I dont think the type of noise is important here

Are you sure the index setting is redundant? What some people call `redundant` some others call `safe` :)

Regarding the multiplication, if you're a numpy freak then you will never be happy with pandas. After all, a pandas df has more info than an ndarray. This doesnt appear to be a matrix product to me. Instead, I think pandas multiplies the series entries with the entire df column based on name matching (breaks down if noise df does have different column names). But if you're unhappy with a (working) implementation for reasons of taste I think its on you to suggest an alternative

Reply:
I think you misunderstood my first point: Uniform is perfectly fine for me and also what I would have chosen. However, you selected `normal` instead, probably by mistake! Correct?

Sorry, did't mean an actual "matrix product" but merely the fact of multiplying matrix (i.e. dataframe) with vector (i.e. series) of some kind. But it's fine, now that the code is already there and you tested it, keep it. It's only that I really hate the way indexes are treated in pandas. I think your statement "be safe" perfectly underpins my point: the fact that one performs seemingly unnecessary extra steps just to be on the "safe side" illustrates how ambiguous the pandas behavior is when it comes to indices. Be honest: you also wouldn't know for certain what would happen if you left the index away without testing 🙃

Reply:
damn, indeed, didnt see the `normal`. fixed

Further upset about pandas though should be directed to the pandas repo and not my PR :D. Again, if you want to propose an implementation that doesnt suffer from these points then we can add it

Reply:
all good 👍🏼  glad we could catch the `normal` bug